src/ParquetMsg.chpl

```diff
@@ -1206,75 +1213,234 @@ module ParquetMsg {
+      var str_idx = (+ scan sections_sizes_str) - sections_sizes_str;
+      var int_idx = (+ scan sections_sizes_int) - sections_sizes_int;
+      var real_idx = (+ scan sections_sizes_real) - sections_sizes_real;
+      var bool_idx = (+ scan sections_sizes_bool) - sections_sizes_bool;
+
+      // populate data based on object and data types
+      forall (i, column, si, ui, ri, bi, oi) in zip(0..#ncols, sym_names, str_idx, int_idx, real_idx, bool_idx, offset_idx) {
         // generate the local c string list of column names
         c_names[i] = my_column_names[i].localize().c_str();
 
         var entry = st.lookup(column);
-
-        // access the dtype of each 
-        var entryDtype = DType.UNDEF;
-        if (entry.isAssignableTo(SymbolEntryType.TypedArraySymEntry)) {
-          entryDtype = (entry: borrowed GenSymEntry).dtype;
-        } else if (entry.isAssignableTo(SymbolEntryType.SegStringSymEntry)) {
-          entryDtype = (entry: borrowed SegStringSymEntry).dtype;
-        } else {
-          throw getErrorWithContext(
-              msg="Unknown SymEntry Type",
-              lineNumber=getLineNumber(), 
-              routineName=getRoutineName(), 
-              moduleName=getModuleName(), 
-              errorClass='ValueError'
-          );
-        }
-        
+        var objType = _identifyObjectType(entry);
+        var entryDtype = _identifyDtype(entry, objType);
         select entryDtype {
           when DType.Int64 {
-            var e = toSymEntry(toGenSymEntry(entry), int);
-            var locDom = e.a.localSubdomain();
-            // set the pointer to the entry array in the list of Pointers
-            if locDom.size > 0 {
-              ptrList[i] = c_ptrTo(e.a[locDom]): c_void_ptr;
+            if objType == ObjType.SEGARRAY {
+              var e: SegArraySymEntry = toSegArraySymEntry(entry, int);
+              var segArray = new SegArray("", e, int);
+              ref sa = segArray;
+              var A = sa.segments.a;
+              const lastOffset = A[A.domain.high];
+              const lastValIdx = sa.values.a.domain.high;
+              const locDom = sa.segments.a.localSubdomain();
+
+              objTypes[i] = ObjType.SEGARRAY: int;
               datatypes[i] = ARROWINT64;
-              sizeList[i] = locDom.size;
+
+              if locDom.size > 0 {
+                var localOffsets = A[locDom];
+                var startValIdx = localOffsets[locDom.low];
+                var endValIdx = if (lastOffset == localOffsets[locDom.high]) then lastValIdx else A[locDom.high + 1] - 1;
+                var valIdxRange = startValIdx..endValIdx;
+                ref olda = sa.values.a;
+                int_vals[ui..#valIdxRange.size] = olda[valIdxRange];
+                ptrList[i] = c_ptrTo(int_vals[ui]): c_void_ptr;
+                sizeList[i] = locDom.size;
+                offset_tracking[oi..#locDom.size] = localOffsets - startValIdx;
+                offsetPtr[i] = c_ptrTo(offset_tracking[oi]);
```

Comment:
`offsetPtr` is always pointing to an array of integers, but `ptrList` may contain a pointer to an `int`, `uint`, `real`, `bool`, or `uint(8)`. Thus, we use the void pointer cast for `ptrList` because we are tracking the dtype and are able to cast it to the correct type in the C++.

Reply:
ah gotcha! I knew `prtList` could contain pointers to different types, but wanted to double check that we didn't want to cast the other since they were both `void**`